File: Section 2.2/Party Lamps/lamps2.java
/*
ID: pranav.19
LANG: JAVA
TASK: lamps
*/

import java.io.*;
import java.util.*;
public class lamps2 {
    static int N;
    static int C;
    static String orig;
    static String finalState = "";
    static Set<String> allSets= new HashSet<String>();
    static Set<String> used = new HashSet<String>();
    public static void main(String[] args) throws IOException {
        runProgram();

    }

    public static String applyCode(String code) {

        String fin = orig;
        for (String i: code.split("")) {
            if (i.equals("4")) {
                fin = button4(fin);
            } else if (i.equals("3")) {
                fin = button3(fin);
            } else if (i.equals("2")) {
                fin = button2(fin);
            } else {
                fin = button1(fin);
            }
        }
        return fin;
    }

    public static void span(String current, int c) {
        // System.out.println(current);
        if (c == 0) {
            span("1", c+1);
            span("2", c+1);
            span("3", c+1);
            span("4", c+1);
            return;
        } else if (used.contains(current)) {
            System.out.println("\t===BREAK===");
            return;
        } else if (c == C) {

            String fin = applyCode(current);

            boolean noFind = true;
            for (int i = 0; i < fin.length(); i++) {
                if (!fin.substring(i, i+1).equals("2")) {
                    noFind = false;
                    break;
                }
            }
            if (noFind) {
                allSets.add(fin);
                return;
            }

            for (int i = 0; i < fin.length(); i++) {
                if (!finalState.substring(i, i+1).equals(fin.substring(i, i+1)) && !finalState.substring(i, i+1).equals("2")) {
                    return;
                }
            }

            // System.out.println("\t===FOUND===");
            allSets.add(fin);
            return;
        } else {
            used.add(current);
            if (current.substring(current.length()-1).equals("4")) {
                span(current+"4", c+1);
            } else if (current.substring(current.length()-1).equals("3")) {
                span(current+"3", c+1);
                span(current+"4", c+1);
            } else if (current.substring(current.length()-1).equals("2")) {
                span(current+"2", c+1);
                span(current+"3", c+1);
                span(current+"4", c+1);
            } else {
                span(current+"1", c+1);
                span(current+"2", c+1);
                span(current+"3", c+1);
                span(current+"4", c+1);
            }
            return;
        }
    }

    public static void runProgram() throws IOException {
        
        System.out.println("STARTED");
        BufferedReader reader = new BufferedReader(new FileReader(new File("lamps.in")));
        PrintWriter writer = new PrintWriter(new File("lamps.out"));

        N = Integer.parseInt(reader.readLine());
        C = Integer.parseInt(reader.readLine());

        String[] temp1 = reader.readLine().split(" ");
        String[] temp2 = reader.readLine().split(" ");

        int[] temp = new int[N];

        for (int i = 0; i < temp.length; i++) {
            temp[i] = 2;
        }
        for (int i = 0; i < temp1.length-1; i++) {
            temp[Integer.parseInt(temp1[i])-1] = 1;
        }
        for (int i = 0; i < temp2.length-1; i++) {
            temp[Integer.parseInt(temp2[i])-1] = 0;
        }

        for (int i: temp) {
            finalState += i;
        }


        String bulbState = "";
        for (int i = 0; i < N; i++) {
            bulbState += "1";
        }
        orig = bulbState;

        long startTime = System.currentTimeMillis();
        span("", 0);
        long endTime = System.currentTimeMillis();
        System.out.println(String.format("Program took %s seconds", (endTime-startTime)/1000.0));

        if (allSets.size() == 0) {
            writer.println("IMPOSSIBLE");
        } else {
            ArrayList<String> sets = new ArrayList<String>();
            for (String i: allSets) {
                sets.add(i);
            }
            Collections.sort(sets);
            for (String i: sets) {
                writer.println(i);
            }
        }
        
        reader.close();
        writer.close();
        
    }

    public static String button1(String s) {
        String n = "";
        for (int i = 0; i < s.length(); i++) {
            n += ""+(Integer.parseInt(s.substring(i, i+1))+1)%2;
        }
        return n;
    }

    public static String button2(String s) {
        String n = "";
        for (int i = 0; i < s.length(); i++) {
            if ((i+1)%2 == 1) {
                n += ""+(Integer.parseInt(s.substring(i, i+1))+1)%2;
            } else {
                n += s.substring(i, i+1);
            }
        }
        return n;
    }

    public static String button3(String s) {
        String n = "";
        for (int i = 0; i < s.length(); i++) {
            if ((i+1)%2 == 0) {
                n += ""+(Integer.parseInt(s.substring(i, i+1))+1)%2;
            } else {
                n += s.substring(i, i+1);
            }
        }
        return n;
    }

    public static String button4(String s) {
        String n = "";
        for (int i = 0; i < s.length(); i++) {
            if (i%3 == 0) {
                n += ""+(Integer.parseInt(s.substring(i, i+1))+1)%2;
            } else {
                n += s.substring(i, i+1);
            }
        }
        return n;
    }

}

File: Section 2.2/Party Lamps/lamps2.py
"""
ID: pranav.19
LANG: PYTHON3
TASK: lamps
"""
import sys
# 0 --> off
# 1 --> on
# 2 --> arb
N = 0
finalC = 0
c = 0
bulbs = ""
finalState = []
with open("lamps.in", "r") as f:
    N = int(f.readline())
    finalC = int(f.readline())
    on = f.readline().split()
    off = f.readline().split()
    for i in range(N):
        bulbs += "1"
        finalState.append("2")

    for i in on[:len(on)-1]:
        finalState[int(i)-1] = "1"
    for i in off[:len(off)-1]:
        finalState[int(i)-1] = "0"
    finalState = "".join(finalState)
# print(*bulbs)
# print("finalState", *finalState)

goods = set([])
used = set([])
sys.setrecursionlimit(40000)

def button1(s):
    return "".join([str((int(i)+1)%2) for i in s])

def button2(s):
    b2 = ""
    for i in range(0, len(s)):
        if i % 2 == 1:
            b2 += str((int(s[i])+1)%2)
        else:
            b2 += s[i]
    return b2

def button3(s):
    b3 = ""
    for i in range(0, len(s)):
        if i % 2 == 0:
            b3 += str((int(s[i])+1)%2)
        else:
            b3 += s[i]
    return b3

def button4(s):
    b4 = ""
    for i in range(0, len(s)):
        if i % 3 == 0:
            b4 += str((int(s[i])+1)%2)
        else:
            b4 += s[i]
    return b4

previousList = {}
currentList = {bulbs}

for c in range(finalC):
    previousList = currentList.copy()
    temp = set([])

    for i in currentList:
        temp.add(button1(i))
        temp.add(button2(i))
        temp.add(button3(i))
        temp.add(button4(i))

    if len(previousList - temp) == 0 and len(temp - previousList) == 0:
        break

    currentList = temp

for state in currentList:
    for i in finalState:
            if i != "2":
                break
    else:
        goods.add(state)
    
    for i in range(len(finalState)):
        # print(len(finalState) == len(bulbState))
        if finalState[i] != "2" and not finalState[i] == state[i]:
            break
    else:
        goods.add(state)
# print("Ended checks")

answers = []
for i in goods:
    answers.append(i)

answers.sort()
# for i in answers:
#     print(i)
with open("lamps.out", "w") as f:
    if len(answers) == 0:
        f.write("IMPOSSIBLE\n")
    else:
        for i in answers:
            f.write("%s\n" % i)